File: cpp/16_3Sum_Closest.cpp
/*
Given an array nums of n integers and an integer target, 
find three integers in nums such that the sum is closest to target. 
Return the sum of the three integers. 
You may assume that each input would have exactly one solution.

Example:
Given array nums = [-1, 2, 1, -4], and target = 1.

The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).
*/
#include <iostream>
#include <vector>

using namespace std;

class Solution {
    public:
        int threeSumClosest(vector<int>& nums, int target) {
            int res = nums[0] + nums[1] + nums[2], len = nums.size(), threeSum;
            sort(nums.begin(), nums.end());
            for(int i = 0; i < len; i++) {
                int j = i + 1, k = len - 1;
                int subTarget = target - nums[i];
                while(j < k) {
                    threeSum = nums[i] + nums[j] + nums[k];
                    if(abs(threeSum - target) < abs(res - target)) {
                        res = threeSum;
                        if(res == target) return res;
                    }
                    threeSum > target ? k-- : j++;
                }
            }
            return res;
        }
};

int main() {
    Solution sol;
    return 0;
}

/*


*/

File: py/16_3Sum_Closest.py
"""
Given an array nums of n integers and an integer target, 
find three integers in nums such that the sum is closest to target. 
Return the sum of the three integers. 
You may assume that each input would have exactly one solution.

Example:
Given array nums = [-1, 2, 1, -4], and target = 1.

The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).
"""
from typing import List

class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        l = len(nums)
        res = nums[0] + nums[1] + nums[2]
        for i in range(l):
            remain = target - nums[i]
            j, k = i + 1, l - 1
            while j < k:
                twoSum = nums[j] + nums[k]
                if twoSum == remain:
                    return target
                if abs(target - res) > abs(twoSum - remain):
                    res = nums[i] + twoSum
                if twoSum > remain:
                    k -= 1
                else:
                    j += 1
        return res
